File: Bronze/2018/US Open/03_family_tree/FamilyTreeI.java
import java.io.BufferedReader;
import java.io.FileReader;
import java.io.PrintWriter;
import java.util.*;

/**
 * @author Red
 * @version 1.0
 */
public class FamilyTreeI {
    static int N;

    public static void main(String[] args) throws Exception {
        BufferedReader r = new BufferedReader(new FileReader("family.in"));
        PrintWriter pw = new PrintWriter("family.out");
        StringTokenizer st = new StringTokenizer(r.readLine());
        N = Integer.parseInt(st.nextToken());
        String cow1 = st.nextToken();
        String cow2 = st.nextToken();
        List<String[]> relations = new ArrayList<>();
        for (int i = 0; i < N; i++) {
            relations.add(r.readLine().trim().split(" "));
        }

        String node = cow1;
        int da = 0;
        while (node != null) {
            if (getAncestorDistance(cow2, node, relations) != -1) {
                break;
            }
            node = rdMother(node, relations);
            da++;
        }

        // cannot rd a common ancestor between a and b
        if (node == null) {
            pw.write("NOT RELATED");
            pw.close();
            System.exit(0);
        }

        int db = getAncestorDistance(cow2, node, relations);
        if (da > 1 && db > 1) {
            pw.write("COUSINS");
            pw.close();
            System.exit(0);
        }
        if (da == 1 && db == 1) {
            pw.write("SIBLINGS");
            pw.close();
            System.exit(0);
        }

        // rd a node with higher generation
        if (da > db) {
            int temp = da;
            da = db;
            db = temp;
            String tempName = cow1;
            cow1 = cow2;
            cow2 = tempName;
        }

        StringBuilder prefix = new StringBuilder("great-".repeat(Math.max(0, db - 2)));
        if (db > 1 && da == 0) {
            prefix.append("grand-");
        }
        if (da == 0) {
            prefix.append("mother");
        } else {
            prefix.append("aunt");
        }

        pw.write(cow1 + " is the " + prefix + " of " + cow2);
        pw.close();
    }

    static String rdMother(String cow, List<String[]> relations) {
        for (String[] r : relations) {
            if (r[1].equals(cow)) {
                return r[0];
            }
        }
        return null;
    }

    static int getAncestorDistance(String sourceCow, String targetCow, List<String[]> relations) {
        int distance = 0;
        String node = sourceCow;
        while (node != null) {
            if (node.equals(targetCow)) {
                return distance;
            }
            node = rdMother(node, relations);
            distance++;
        }
        return -1;
    }
}

File: Bronze/2018/US Open/03_family_tree/family_tree_i.cpp
#include <iostream>
#include <fstream>
#include <vector>
#include <algorithm>

using namespace std;

string findMother(string cow, const vector<vector<string>>& relations) {
    for (const auto& r : relations) {
        if (r[1] == cow) {
            return r[0];
        }
    }
    return "";
}

int getAncestorDistance(const string& sourceCow, const string& targetCow, const vector<vector<string>>& relations) {
    int distance = 0;
    string node = sourceCow;
    while (!node.empty()) {
        if (node == targetCow) {
            return distance;
        }
        node = findMother(node, relations);
        distance++;
    }
    return -1;
}

int main() {
    ifstream fin("family.in");
    ofstream fout("family.out");

    int N;
    fin >> N;
    string name_a, name_b;
    fin >> name_a >> name_b;

    vector<vector<string>> relations(N, vector<string>(2));
    for (int i = 0; i < N; ++i) {
        fin >> relations[i][0] >> relations[i][1];
    }

    string node = name_a;
    int da = 0;
    while (!node.empty()) {
        if (getAncestorDistance(name_b, node, relations) != -1) {
            break;
        }
        node = findMother(node, relations);
        da++;
    }

    // cannot find a common ancestor between a and b
    if (node.empty()) {
        fout << "NOT RELATED";
        return 0;
    }

    int db = getAncestorDistance(name_b, node, relations);
    if (da > 1 && db > 1) {
        fout << "COUSINS";
        return 0;
    }
    if (da == 1 && db == 1) {
        fout << "SIBLINGS";
        return 0;
    }

    // find a node with a higher generation
    if (da > db) {
        swap(da, db);
        swap(name_a, name_b);
    }

    string prefix = "";
    int repeat = max(0, db - 2);
    for (int i = 0; i < repeat; ++i) {
        prefix += "great-";
    }
    if (db > 1 && da == 0) {
        prefix += "grand-";
    }
    if (da == 0) {
        prefix += "mother";
    } else {
        prefix += "aunt";
    }

    fout << name_a << " is the " << prefix << " of " << name_b;
    return 0;
}
